### ISSUE TITLE
feature: Allow adding arbitrary claim in transform

### ISSUE THREAD

**Issue by roblabla** [greenpau/caddy-auth-portal#151]
**Describe the issue**

I'd like to use the caddy auth portal as an authentication frontend for CouchDB, allowing static configuration of CouchDB users in a simple way. [CouchDB JWT support](https://docs.couchdb.org/en/stable/api/server/authn.html?highlight=jwt#jwt-authentication) allows this in a fairly straightforward way.

CouchDB uses a custom claim, `_couchdb.roles`, to configure the roles of each user. This is currently not possible to express as a transform in the Caddy Auth Portal, as the transforms appear to be limited to known, hardcoded claims.

Ideally, I'd like to write this in my caddyfile:

```
:443 {
	route /auth* {
		authp {
                       # Elided backend config - uses github oauth2.
			transform user {
				exact match sub github.com/roblabla
				action add roles users
				action add _couchdb.roles _admin
			}
			crypto key sign-verify whatever
		}
	}

	route /* {
		handle /couchdb {
			uri strip_prefix /couchdb
			reverse_proxy http://localhost:5984
		}
	}
}
```

and have it result in my JWT token automatically getting the `_couchdb.roles` claim set to the _admin value.

**Version Information**

Provide output of `caddy list-modules -versions | grep auth` below:

```
http.authentication.hashes.bcrypt v2.4.3
http.authentication.hashes.scrypt v2.4.3
http.authentication.providers.http_basic v2.4.3
http.handlers.authentication v2.4.3
http.authentication.providers.jwt v1.3.13
http.handlers.authp v1.4.17
```

**Comment by greenpau** [comment]
@roblabla , that is doable. Will look into it next week!

**Comment by roblabla** [comment]
FWIW, the `main` branch of couchdb has an option to override the claim used for the couchdb roles, so I managed to hack around this issue by setting its roles claim to `roles`. Still, I'm pretty sure there are other software that would benefit from the ability to set arbitrary claims.

**Comment by andoks** [comment]
Adding custom claims with the possibility to inject dynamic values like for the [inject header](https://authp.github.io/docs/authorize/headers#custom-headers) functionality could also be useful when using auth-portal/authorize with services like Hasura ([Hasura JWT Spec](https://hasura.io/docs/latest/graphql/core/auth/authentication/jwt.html#the-spec)), e.g `action add "https://hasura.io/jwt/claims.x-hasura-user-id" from email`.

**Comment by greenpau** [comment]
> inject header functionality could also be useful when using auth-portal/authorize with services like Hasura

@andoks , will try finding time working on it. Great example! 👍 Thank you for the link! 👏

**Comment by andoks** [comment]
@greenpau, happy to "help" :smile: 

If you end up implementing dynamic claim sources (like the users' email etc), could you also please add the user-id as a "source" field too? I would like to use the users' ID (extracted from the JWT) in the future if possible.

**Comment by greenpau** [comment]
> If you end up implementing dynamic claim sources (like the users' email etc), could you also please add the user-id as a "source" field too? I would like to use the users' ID (extracted from the JWT) in the future if possible.

@andoks , please open a separate issue for this. I am going to close this issue today.

**Comment by greenpau** [comment]
@roblabla , it took a while to get to this 😃 You probably moved on ... Nevertheless, I liked the issue and always wanted to get back to it. Now, that I am moving towards PASETO, was a good time to address it.

Released [v1.4.34](https://github.com/greenpau/caddy-auth-portal/releases/tag/v1.4.34)

```
            transform user {
                match origin local
                action add _couchdb.roles _admin as string list
            }
```

![image](https://user-images.githubusercontent.com/3826416/145342332-0ddaf595-adb1-4cef-84d0-5d452e535109.png)
